### PR TITLE
DM-43800: Rewrite Felis user guide

### PR DESCRIPTION
**Pull request by JeremyMcCormick**
This represents a complete rewrite of the Felis user guide at [felis.lsst.io](https://felis.lsst.io).

### CODE REVIEW

File: docs/user-guide/validation.rst
Comment (andy-slac) [inline docs/user-guide/validation.rst:16]: JSON?

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:5]: Hmm, "create a database" sounds incomplete, in my mental model this translates into `CREATE DATABASE foo;` which is probably not what Felis does. Maybe say "create a database schema"?

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:43]: Duplicate `schema.yaml`?

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:35]: Hmm, not clear why this is dry-run? Just because `mysql://` is not a valid URL? This may be confusing, I think a required explicit flag would be safer.

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:37]: Generate and do what? Dump it to standard output?

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:46]: Maybe, to reduce confusion, dry run should only require dialect name instead of full URL?

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:71]: Hmm, this does not parse. Postgres has databases and schemas inside databases. SQLAlchemy URLS specify database name as a first part of the path after host:port. So I'm confused is this is actually the database name in which a new schema will be created, or this specifies schema name and database name comes from Felis schema name (which sounds odd).

File: docs/user-guide/databases.rst
Comment (andy-slac) [inline docs/user-guide/databases.rst:66]: Driver is optional in SQLAlchemy, may be worth mentioning that.

File: docs/user-guide/examples.rst
Comment (andy-slac) [inline docs/user-guide/examples.rst:36]: Could you explain the use of `#` in `@id`?

**Review by andy-slac — COMMENTED** [review]
Few comments/questions after quick reading of the new user guide. Looks great in general, maybe a bit too terse in places.

File: docs/user-guide/installation.rst
Comment (JeremyMcCormick) [reply docs/user-guide/installation.rst:25]: This is a typo that I will fix. Thank you!

File: docs/user-guide/model.rst
Comment (JeremyMcCormick) [reply docs/user-guide/model.rst:5]: I think according to most style guides, this should be fine as is.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:5]: I will reword this and clarify. Agree that it is somewhat unclear.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:37]: Print to the console - I will clarify this.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:46]: I see your point but changing this is beyond the scope of this PR, since it only deals with documentation.

Being able to execute `felis create --engine-url mysql://` and have dry run mode be enabled automatically is essentially a convenience feature.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:71]: I will clarify for since the syntax is different than with MySQL. You're right that target database has to be created beforehand and provided in the URL for Postgres.

Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:71]: In MySQL, the database name is generally left out of the URL and it will come from the file automatically. Or it can be overridden using a CLI option.

For PostgreSQL, the database name must be included in the URL, and the schema is created within the target database, using the name within the schema or that from the CLI option.

I agree that it is a bit confusing since the URL is different depending on the target database variant (MySQL vs PostgreSQL). I do mention this several times in the documentation though.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:66]: Okay, I will mention this.

File: docs/user-guide/examples.rst
Comment (JeremyMcCormick) [reply docs/user-guide/examples.rst:36]: I can't really explain it other than "Someone did it initially and everyone else copied it."

I guess it means "in this file" like an HTML link? Then if you were combining names the references could look like "myschema#myid."

I'm not entirely sure though. There's no requirement that the IDs have "#" in them. It is just a convention.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:46]: > dry run should only require dialect name instead of full URL?

That is the case currently. You can use `mysql://` as the engine URL, for instance.

This text is just indicating that if you provide a full URL in dry run mode, everything but the dialect will be ignored by SQLAlchemy (due to how mock engine works).

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:35]: > I think a required explicit flag would be safer.

Possibly - it is something we could discuss but not going to change this behavior within this PR, since it is a documentation update, only.

File: docs/user-guide/databases.rst
Comment (JeremyMcCormick) [reply docs/user-guide/databases.rst:80]: This is misworded for Postgres. I will fix it.

File: docs/index.rst
Comment (gpdf) [inline docs/index.rst:5]: Add:
"It covers not only the metadata required to create database schemas, but also a wide range of additional metadata associated with IVOA standards and other applications."

File: docs/user-guide/intro.rst
Comment (gpdf) [inline docs/user-guide/intro.rst:19]: Add:
"Compatible TAP services can use this data to populate their `/tables` output as well."

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:8]: ... extra metadata to these elements, like units ...

(This avoids a grammatical implication that the "extra metadata" applies only to columns.)

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:64]: ... order the tables in public presentations of the schema.

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:91]: ... order the columns in public presentations of the table.

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:92]: ... is a "principal column" in a TAP table; principal columns represent a subset to be highlighted or used as a default in public presentations and query-builders.

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:93]: ... whether this column is a representation of an element of an IVOA-standard data model.

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:96]: ... xtype, if any, for this column.

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:86]: Please put these in the order "ivoa:unit", "fits:tunit", and "ivoa:ucd", to keep the units together.

File: docs/user-guide/datatypes.rst
Comment (gpdf) [inline docs/user-guide/datatypes.rst:19]: We need a footnote on `byte` here and in the next table to explain the issue with signedness.

File: docs/user-guide/datatypes.rst
Comment (gpdf) [inline docs/user-guide/datatypes.rst:35]: ... in VOTable, and is not usable for UTF-8 strings.

File: docs/user-guide/tap.rst
Comment (gpdf) [inline docs/user-guide/tap.rst:47]: Note that standards-conformant TAP service will generally also make the `TAP_SCHEMA` data available at a `/tables` endpoint as well.

**Review by gpdf — COMMENTED** [review]
A number of small changes requested.

I'm using "Comment" just so that you don't have to wait for another review.

**Review by gpdf — APPROVED** [review]
Changing to "Approve" on the understanding that the previously requested changes will be made today.

File: docs/user-guide/datatypes.rst
Comment (JeremyMcCormick) [reply docs/user-guide/datatypes.rst:19]: Could you provide text for this? I'm not sure what you'd like to say about it.

Comment (gpdf) [reply docs/user-guide/datatypes.rst]: "The Felis type `byte` currently maps to signed 8-bit integral types in the supported database back ends.  VOTable, and therefore TAP_SCHEMA, only has an **unsigned** 8-bit type.  The Felis mapping is under review at this time; in the mean time, if you are intentionally storing signed byte values in a database with a Felis-derived schema, we recommend the use of a VOTable type override in Felis, `votable:datatype: short` to ensure that the signedness of the values is preserved in TAP output."

File: docs/user-guide/model.rst
Comment (gpdf) [inline docs/user-guide/model.rst:8]: comma after "elements"

**Comment by JeremyMcCormick** [comment]
@gpdf All suggested changes incorporated.

File: docs/user-guide/model.rst
Comment (JeremyMcCormick) [reply docs/user-guide/model.rst:8]: > comma after "elements"

Missed this, but will incorporate it into a user branch with additional small fixes if I find any other grammar issues, etc.

**Pull request merged**
